src/Compilers/CSharp/Portable/CodeGen/EmitAddress.cs

```diff
@@ -64,6 +64,13 @@ private LocalDefinition EmitAddress(BoundExpression expression, AddressKind addr
 
                     if (expression.Type.IsValueType)
                     {
+
+                        if (addressKind == AddressKind.Writeable && _method.IsReadOnly)
+                        {
+                            // a readonly method is calling a non-readonly method, therefore we need to copy 'this'
+                            goto default;
```

Comment:
will we allow something like `this.x++`  ?

Reply:
Ah, I see, we definitely will not :-)

---
In reply to: [263628279](https://github.com/dotnet/roslyn/pull/33954#discussion_r263628279) [](ancestors = 263628279)

Reply:
I believe if `x` is a value type it will be a compile error, because `this` will be treated as readonly in the context of the readonly method.

Reply:
Actually, seems like any `++` operator will fail, because it appears to assign the return value to the operand.